src/base/dof_map_constraints.C

```diff
@@ -2165,6 +2166,65 @@ void DofMap::enforce_constraints_exactly (const System & system,
   v->close();
 }
 
+void DofMap::enforce_constraints_on_residual (const NonlinearImplicitSystem & system,
+                                              NumericVector<Number> * rhs,
+                                              NumericVector<Number> const * solution) const
+{
+  parallel_object_only();
+
+  if (!this->n_constrained_dofs())
+    return;
+
+  if (!rhs)
+    rhs = system.rhs;
+  if (!solution)
+    solution = system.solution.get();
+
+  libmesh_assert_equal_to (this, &(system.get_dof_map()));
+
+  for (const auto & pr : _dof_constraints)
+    {
+      dof_id_type constrained_dof = pr.first;
+      if (!this->local_index(constrained_dof))
+        continue;
+
+      const DofConstraintRow constraint_row = pr.second;
+
+      Number exact_value = 0;
+      for (const auto & j : constraint_row)
+        exact_value += j.second * (*solution)(j.first);
+      exact_value -= (*solution)(constrained_dof);
```

Comment:
No _primal_constraint_values here?  Supporting only homogeneous constraints would be a step up from supporting nothing, but we'd at least want to scream and die if anyone ever tried to use this with heterogeneous constraints, not silently return wrong results.

Reply:
I'm not too knowledgeable about the "primal constraint" terminology. Would the primal constraints be the b vector in `u = C x + b` (using your syntax from [here](https://github.com/libMesh/libmesh/issues/1985#issuecomment-451596272))

Reply:
@roystgnr ?

Reply:
Sorry, missed that comment the first time.  Yes, _primal_constraint_values would be the sparse b in that equation.  (well, 99.9% sure it's b; otherwise I've misremembered the sign convention and it's -b)

Reply:
Ok @roystgnr I added possible contributions from `_primal_constraint_values`